### ISSUE TITLE
The new GAN version doesn't seem to work

### ISSUE THREAD

**Issue by Nelthirion** [shaoanlu/faceswap-GAN#11]
I've been training the new version for 10000 + 6000 iterations and the output (from the `show_g` function) doesn't even start to change... I saw some outputs where the network tried to mimic PersonB but at the end of the day the three columns `[test_A, path_A(test_A), path_B(test_A)]` all look the same as `test_A` and vice versa...

Not only that, when I tried turning `use_mixup` to `False` I had this error about different number of channels here:

`number of input channels does not match corresponding dimension of filter, 3 != 6`
`output_real = netD(real)  # positive 

It seems we have to manually change `nc_D_inp` to 3 instead of 6.

Keep up the good work...

**Comment by shaoanlu** [comment]
Ahh, I didn't expect anyone will modify `use_mixup` to `True`. So I did not change the code in that part. Thanks for pointing it out. 

Regarding failed learning, what is the output mask looks like after you trained it for 16k iters.? Does the model learn good segmentation masks (at least close to mine)?

Also, you can modify the code in section "Start Training" where
```python
# get new batch of images and generate results for visualization
...
showG(tA, tB, path_A, path_B)
showG_mask(tA, tB, path_mask_A, path_mask_B)  
```
to be
``` python
showG(wA, wB, path_A, path_B)
showG_mask(wA, wB, path_mask_A, path_mask_B)  
```
, and see if the model can recover warped images back to unwarped ones successfully.

I'm retraining a model from scratch using the v2_train script now to see if anything goes wrong. So back to you later.

**Comment by Nelthirion** [comment]
[Here's](https://imgur.com/a/NSZEe) what the output was at 100 iterations.

And [here's](https://imgur.com/a/5ghph) what the output looks like after 10000 + 6000 Training (using the different code you provided for each training range).

Funny thing is that the 16000th iteration generated images doesn't really have a difference than the first one! The mask has surely changed throughout training but the generated images have little change (can't really say they didn't change).

**Comment by shaoanlu** [comment]
I think I've figured out what is missing: the regularization of alpha mask, which is commented out in the releasing jupyter.
```python
loss_GA += 1e-2 * K.mean(K.abs(mask_A))
loss_GB += 1e-2 * K.mean(K.abs(mask_B))
```
I trained my original model with the regularization for 6.6k iters and then gradually decreased (3e-3 for ~4k following by 1e-3 for ~5k) and finally removed it. I did not find much difference on output mask before/after removing regularization at that time. So I considered it trivial (apparently it backfires).

**Comment by Nelthirion** [comment]
Hmm... Thanks for the new info... I'm gonna try it out...

By the way, can you perhaps use a moving average or a decaying algorithm to automate the decay of regularization?

**Comment by shaoanlu** [comment]
I suggest using mask regularization as `3e-3 * K.mean(K.abs(mask))` and don't change loss function at all. This will give result that is good enough.

**Comment by Nelthirion** [comment]
I'm training it with perceptual loss now... 

Training without it (and with all the same settings as your new modified code) gave bad results even after 27000 iterations... [Here's](https://imgur.com/a/SQBh7) the 27000th iteration without perceptual loss.

I have narrowed it to 3 possibilities:

1. Bad Face Datasets: Input_A dataset is from a 10 minute video and Input_B dataset is from different interviews from different angles (both have high quality and low quality images). I wish you could propose a way to get your training dataset just for the sake of making sure that our training is done right.

2. No perceptual loss: I didn't use perceptual loss before because the batch_size was more than my VRAM could chew. now that it's 8 batches I can train it with perceptual loss. guess we'll see after 20k+ iterations.

3. Different versions of Keras, Tensorflow and VGGFace: I have Keras v2.0.6, Tensorflow v1.4.0 and the latest VGGFace (just downloaded it today with pip). I'm not sure if this really can make problems. but for example your code wouldn't work with VGGFace's ResNet since I had problems like [this](https://github.com/rcmalli/keras-vggface/issues/14) and had to change the VGGFace code a bit.

Thanks for the updates and I'll be sure keep you informed...

**Comment by Nelthirion** [comment]
OK... Here's my take on the training...

It seems you have to have a LOT of useful images from both parties (Mainly the second one) for the training to progress smoothly...

Since I didn't have enough data, My training (even after 30000 iterations) always end up with the mask occluding the center of the eyes and having monstrous outputs!

Your output mask example doesn't seem to have this occlusion on the eyes... I'm going to put the blame on my small dataset (nearly 500 images)...

**Comment by shaoanlu** [comment]
Hmm, maybe alpha mask is hard for model to learn. Thanks for letting me know.
